### 1. Wczytaj dane z pliku heart.csv i zapoznaj się z nimi. Wygeneruj statystyki.

In [130]:
import pandas as pd

pd.set_option('display.float_format', '{:0.2f}'.format)
data = pd.read_csv('heart.csv')

data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0     2.30    0   
1   37    1   2     130   250    0        1       187     0     3.50    0   
2   41    0   1     130   204    0        0       172     0     1.40    2   
3   56    1   1     120   236    0        1       178     0     0.80    2   
4   57    0   0     120   354    0        1       163     1     0.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [131]:
data.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.37   0.68   0.97  131.62 246.26   0.15     0.53    149.65   0.33   
std     9.08   0.47   1.03   17.54  51.83   0.36     0.53     22.91   0.47   
min    29.00   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    47.50   0.00   0.00  120.00 211.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 240.00   0.00     1.00    153.00   0.00   
75%    61.00   1.00   2.00  140.00 274.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [132]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [133]:
data.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

### 2. Całkowicie losowo wygeneruj brakujące dane tak, aby w kolumnach age i chol brakowało po około 10% wartości.


In [134]:
import numpy as np
import random 
rand_data = data.copy()

missing = 0.1
columns = ['age', 'chol']

for col in columns:
    i = 0
    while i < len(rand_data):
        p = random.random()
        if(p<=missing+0.0001):
            rand_data.loc[i, col] = np.nan
        i+=1

In [135]:
rand_data.isnull().sum()

age         25
sex          0
cp           0
trtbps       0
chol        18
fbs          0
restecg      0
thalachh     0
exng         0
oldpeak      0
slp          0
caa          0
thall        0
output       0
dtype: int64

In [136]:
rand_data.head(10)

age  sex  cp  trtbps   chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0 63.00    1   3     145 233.00    1        0       150     0     2.30    0   
1 37.00    1   2     130 250.00    0        1       187     0     3.50    0   
2 41.00    0   1     130 204.00    0        0       172     0     1.40    2   
3 56.00    1   1     120 236.00    0        1       178     0     0.80    2   
4 57.00    0   0     120 354.00    0        1       163     1     0.60    2   
5 57.00    1   0     140 192.00    0        1       148     0     0.40    1   
6 56.00    0   1     140 294.00    0        0       153     0     1.30    1   
7 44.00    1   1     120 263.00    0        1       173     0     0.00    2   
8 52.00    1   2     172 199.00    1        1       162     0     0.50    2   
9 57.00    1   2     150 168.00    0        1       174     0     1.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

### 3. Usuń rekordy zawierające jakiekolwiek brakujące wartości.

In [137]:
drop_rand_data = rand_data.dropna()

print(len(drop_rand_data))

262


In [138]:
drop_rand_data.isnull().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

In [139]:
drop_rand_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 262 entries, 0 to 301
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       262 non-null    float64
 1   sex       262 non-null    int64  
 2   cp        262 non-null    int64  
 3   trtbps    262 non-null    int64  
 4   chol      262 non-null    float64
 5   fbs       262 non-null    int64  
 6   restecg   262 non-null    int64  
 7   thalachh  262 non-null    int64  
 8   exng      262 non-null    int64  
 9   oldpeak   262 non-null    float64
 10  slp       262 non-null    int64  
 11  caa       262 non-null    int64  
 12  thall     262 non-null    int64  
 13  output    262 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 30.7 KB


### 4. Dokonaj imputacji metodą średniej.

In [140]:
mean_imputation = rand_data.fillna(rand_data.mean()) # wstawiamy średnia w miejsce NaN

mean_imputation.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.32   0.68   0.97  131.62 246.54   0.15     0.53    149.65   0.33   
std     8.47   0.47   1.03   17.54  50.51   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    49.00   0.00   0.00  120.00 212.00   0.00     0.00    133.50   0.00   
50%    54.32   1.00   1.00  130.00 244.00   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 273.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [141]:
mean_imputation.head(10)

age  sex  cp  trtbps   chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0 63.00    1   3     145 233.00    1        0       150     0     2.30    0   
1 37.00    1   2     130 250.00    0        1       187     0     3.50    0   
2 41.00    0   1     130 204.00    0        0       172     0     1.40    2   
3 56.00    1   1     120 236.00    0        1       178     0     0.80    2   
4 57.00    0   0     120 354.00    0        1       163     1     0.60    2   
5 57.00    1   0     140 192.00    0        1       148     0     0.40    1   
6 56.00    0   1     140 294.00    0        0       153     0     1.30    1   
7 44.00    1   1     120 263.00    0        1       173     0     0.00    2   
8 52.00    1   2     172 199.00    1        1       162     0     0.50    2   
9 57.00    1   2     150 168.00    0        1       174     0     1.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

### 5. Porównaj otrzymane rezultaty z oryginalnymi danymi.


In [142]:
# data['age'].compare(mean_imputation['age'])
data['chol'].compare(mean_imputation['chol'])

self  other
12  266.00 246.54
17  226.00 246.54
19  239.00 246.54
35  177.00 246.54
109 254.00 246.54
111 126.00 246.54
130 201.00 246.54
139 263.00 246.54
145 245.00 246.54
153 278.00 246.54
158 220.00 246.54
177 335.00 246.54
183 230.00 246.54
211 260.00 246.54
234 322.00 246.54
254 273.00 246.54
281 204.00 246.54
302 236.00 246.54

### 6. Dokonaj imputacji metodami uzupełniania braków poprzednią wartością, interpolacji, MICE i k-najbliższych sąsiadów.


#### *Uzupełnianie poprzednią wartością (Forward Fill)*

In [ ]:
forward_fill = rand_data.copy().ffill()#.dropna()
forward_fill = forward_fill.fillna(forward_fill.mean) # (jesli np. w pierwszym rekordzie bedzie brak to forward fill zostawi NaN, nie wypelni go)

forward_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.35   0.68   0.97  131.62 246.78   0.15     0.53    149.65   0.33   
std     8.73   0.47   1.03   17.54  52.23   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    48.00   0.00   0.00  120.00 211.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 241.00   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 275.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

#### *Uzupełnianie metodami interpolacji*

In [144]:
interpolation_fill = rand_data.fillna(rand_data.interpolate(method = 'linear'))

interpolation_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.34   0.68   0.97  131.62 246.72   0.15     0.53    149.65   0.33   
std     8.56   0.47   1.03   17.54  51.62   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    48.00   0.00   0.00  120.00 211.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 242.00   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 275.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [145]:
interpolation_fill2 = rand_data.fillna(rand_data.interpolate(method = 'polynomial', order = 2))

interpolation_fill2.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 302.00 303.00   303.00    303.00 303.00   
mean   54.35   0.68   0.97  131.62 247.38   0.15     0.53    149.65   0.33   
std     8.63   0.47   1.03   17.54  51.84   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    48.00   0.00   0.00  120.00 211.25   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 241.50   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 277.00   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

#### *Uzupelnianie MICE*

In [146]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imputer = IterativeImputer(max_iter = 15)
mice_fill = pd.DataFrame(imputer.fit_transform(rand_data), columns=rand_data.columns, index = rand_data.index)

mice_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.33   0.68   0.97  131.62 246.86   0.15     0.53    149.65   0.33   
std     8.58   0.47   1.03   17.54  50.55   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    48.01   0.00   0.00  120.00 212.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 244.00   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 273.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

#### *Uzupelnianie metodą k-nn*

In [147]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=5)
knn_fill = pd.DataFrame(imputer.fit_transform(rand_data), columns=rand_data.columns, index = rand_data.index)

knn_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.36   0.68   0.97  131.62 247.23   0.15     0.53    149.65   0.33   
std     8.58   0.47   1.03   17.54  50.80   0.36     0.53     22.91   0.47   
min    34.00   0.00   0.00   94.00 131.00   0.00     0.00     71.00   0.00   
25%    48.50   0.00   0.00  120.00 212.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 243.00   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 274.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

### 7. Dla każdej z metod policz najmniejszy i największy błąd, a także miary MAE i RMSE.

In [148]:
imputation_methods = {
    "Średnia": mean_imputation,
    "Forward Fill": forward_fill,
    "Interpolacja (Liniowa)": interpolation_fill,
    "Interpolacja (Wielom. 2)": interpolation_fill2,
    "MICE": mice_fill,
    "k-NN (k=5)": knn_fill
}

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

columns_to_impute = ['age', 'chol']
results = {}

for col in columns_to_impute:
    original_missing_values = data[col][rand_data[col].isna()]
    col_results = {}
    
    for method_name, imputed_data in imputation_methods.items():
        imputed_values = imputed_data[col][rand_data[col].isna()]
        errors = np.abs(original_missing_values - imputed_values)
        min_error = errors.min() #min_blad
        max_error = errors.max() #max_blad
        mae = mean_absolute_error(original_missing_values, imputed_values) #MAE
        rmse = np.sqrt(mean_squared_error(original_missing_values, imputed_values)) #RMSE
        col_results[method_name] = {
            "Najmniejszy Błąd": min_error,
            "Największy Błąd": max_error,
            "MAE": mae,
            "RMSE": rmse
        }
    
    results[col] = pd.DataFrame.from_dict(col_results, orient='index')


❌ ERROR in Imputation for: chol using method: Interpolacja (Wielom. 2)
   Imputed values still contain 1 NaNs.
   Skipping error calculation for this method/column.


In [150]:
print(results['age'])

                          Najmniejszy Błąd  Największy Błąd   MAE  RMSE
Średnia                               0.32            25.32  9.98 11.43
Forward Fill                          5.00            23.00 12.40 13.47
Interpolacja (Liniowa)                0.50            22.67 10.61 12.21
Interpolacja (Wielom. 2)              0.42            28.35 11.90 13.88
MICE                                  0.01            16.46  7.46  9.14
k-NN (k=5)                            0.40            19.00  7.88  9.68


In [151]:
print(results['chol'])

                        Najmniejszy Błąd  Największy Błąd   MAE  RMSE
Średnia                             1.54           120.54 35.56 47.68
Forward Fill                        8.00           199.00 65.94 81.68
Interpolacja (Liniowa)              1.50           193.00 58.14 74.09
MICE                               10.69           131.69 36.72 47.97
k-NN (k=5)                          0.80           110.40 37.84 47.09


### 8. Utwórz wykres pokazujący zależność między poziomem cholesterolu a wiekiem pacjenta dla danych oryginalnych i po przeprowadzonych imputacjach.

### 9. Przeprowadź analogiczne analizy dla 30% brakujących wartości.

In [152]:
# Wystarczy zmienic 0.1 na 0.3, nic nie przepisywac

### 10. Zbadaj hipotezę, że średni poziom cholesterolu wynosi 250.

### 11. Sprawdź, czy na podstawie poziomu cholesterolu jesteś w stanie rozpoznać płeć pacjenta.